### PR TITLE
magento-engcom/community-portal#211 Implement rankBySkill Handler

### PR DESCRIPTION
**Pull request by XiyaMagento**
magento-engcom/community-portal#211 Implement rankBySkill Handler

- [x] Implement rankBySkill handler to get the users ranked by the degree of matching according to the skills required by the project
- [x] The handler should also rank the projects in terms of degree of match for each user
- [x] The handler should then store the relevant results to the respective tables
- [x] Schedule the handler to run on a routine basis
- [x] Modify getProjectCards and getProjectDetails so that the information on ranked users is not returned to regular users
- [ ] Create handlers and frontend for users to add their skills and for admins to add required skills for projects (alternatively, infer the skills required based on which parts of the Magento system the project is about)
- [ ] Add data flows to add skills that users have and skills that project require to the skills table in the relevant handlers
- [ ] Create frontend to display projects with highest degree of skill match

I am currently scheduling the handler to run every minute. This allows us to test the results more easily. But we should change the rate before shipping to staging or production.

### CODE REVIEW

File: functions/src/algorithms/rankBySkill.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  return (count: number, key: string, map: Map<string, number>) => {
+    divide(map, key, dividerDict.get(key));
+  };
+}
+
+function topValues(map: Map<string, number>, k: number) {
+  return [...map.entries()].sort(([k1, v1], [k2, v2]) => v2 - v1).slice(0, k);
+}
+
+function rankBySkill(data: DataInterface[]) {
```
Comment (vrann) [inline functions/src/algorithms/rankBySkill.ts]: please add description on how rank by skill algorithm works, either in code or on wiki

Comment (XiyaMagento) [reply functions/src/algorithms/rankBySkill.ts]: Okay! I'll update this comment when I finish adding the description

Comment (XiyaMagento) [reply functions/src/algorithms/rankBySkill.ts]: I have finished writing documentation for the algorithm: https://github.com/magento-engcom/community-portal/wiki/Skill-Matching

**Comment by XiyaMagento** [comment]
The test if failing because of schema change. Will look into that

**Comment by XiyaMagento** [comment]
Responding to the comment: "Should this functionality also have a frontend part to see best suited projects?"

Yes, I agree with that. Should I add it to the issue description?

**Pull request closed without merging**